apps/builder/app/routes/rest.domain.ts

```diff
@@ -0,0 +1,78 @@
+import type { ActionArgs } from "@remix-run/node";
+import { zfd } from "zod-form-data";
+import { prisma } from "@webstudio-is/prisma-client";
+import { trpcClient } from "~/services/trpc.server";
+
+const schema = zfd.formData({
+  projectId: zfd.text(),
+  userId: zfd.text(),
+  customDomain: zfd.text(),
+  customCname: zfd.text(),
+});
+
+async function addDomain({
+  customDomain,
+  projectId,
+  userId,
+}: {
+  customDomain: string;
+  projectId: string;
+  userId: string;
+}) {
+  return await prisma.customDomain.create({
+    data: {
+      domain: customDomain,
+      projectId,
+      userId,
+    },
+  });
+}
+export const action = async ({ request }: ActionArgs) => {
+  const { customDomain, customCname, projectId, userId } = schema.parse(
+    await request.formData()
+  );
+
+  const customDomainFound = await prisma.customDomain.findFirst({
+    where: {
+      domain: customDomain,
+    },
+  });
+  if (customDomainFound) {
+    return {
+      success: false,
+      errors: "Domain already exists! Contact us for support.",
+    };
+  }
+
+  const foundCname = await trpcClient.customDomain.verifyCNAME.mutate({
+    customCname,
+    customDomain,
+  });
+
+  if (!foundCname.success) {
+    return {
+      success: false,
+      errors: foundCname?.error?.message ?? "CNAME not set!",
+    };
+  }
+  const addedCfDomain =
+    await trpcClient.customDomain.addCloudflareCustomHostname.mutate({
+      customDomain,
+    });
+  if (addedCfDomain.success) {
+    await addDomain({
+      customDomain,
+      projectId,
+      userId,
+    });
+    return {
+      success: true,
+    };
+  }
+  // Should handle this error better
+  // CF duplicate domain error or other unknown error
+  return {
```

Comment:
Should be not an issue if it is already created. From my understanding, all custom domain has the same fallback origin value. So it can already be created by the same user or other user